File: api/app/routers/tasks.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.task import TaskCreate, TaskRead, TaskStatusUpdate, TaskUpdate
from app.services.task_service import (
    create_task,
    delete_task,
    get_task,
    get_tasks,
    update_task,
    update_task_status,
)

router = APIRouter()


@router.get("/tasks", response_model=list[TaskRead])
async def list_tasks(
    product_id: uuid.UUID = Query(...),
    db: AsyncSession = Depends(get_db),
):
    return await get_tasks(db, product_id)


@router.post("/tasks", response_model=TaskRead, status_code=201)
async def create_task_endpoint(
    data: TaskCreate, db: AsyncSession = Depends(get_db)
):
    return await create_task(db, data)


@router.get("/tasks/{task_id}", response_model=TaskRead)
async def get_task_endpoint(
    task_id: uuid.UUID, db: AsyncSession = Depends(get_db)
):
    return await get_task(db, task_id)


@router.put("/tasks/{task_id}", response_model=TaskRead)
async def update_task_endpoint(
    task_id: uuid.UUID,
    data: TaskUpdate,
    db: AsyncSession = Depends(get_db),
):
    return await update_task(db, task_id, data)


@router.delete("/tasks/{task_id}", status_code=204)
async def delete_task_endpoint(
    task_id: uuid.UUID, db: AsyncSession = Depends(get_db)
):
    await delete_task(db, task_id)


@router.patch("/tasks/{task_id}/status", response_model=TaskRead)
async def update_task_status_endpoint(
    task_id: uuid.UUID,
    data: TaskStatusUpdate,
    db: AsyncSession = Depends(get_db),
):
    return await update_task_status(db, task_id, data.status)

File: web/src/api/tasks.ts
import { fetchApi } from './client';
import type { Task, TaskCreate, TaskUpdate, TaskStatus } from '../types';

export function getTasks(productId: string): Promise<Task[]> {
  return fetchApi<Task[]>(`/products/${productId}/tasks`);
}

export function getTask(id: string): Promise<Task> {
  return fetchApi<Task>(`/tasks/${id}`);
}

export function createTask(data: TaskCreate): Promise<Task> {
  return fetchApi<Task>('/tasks', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export function updateTask(id: string, data: TaskUpdate): Promise<Task> {
  return fetchApi<Task>(`/tasks/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export function deleteTask(id: string): Promise<void> {
  return fetchApi<void>(`/tasks/${id}`, { method: 'DELETE' });
}

export function updateTaskStatus(id: string, status: TaskStatus): Promise<Task> {
  return fetchApi<Task>(`/tasks/${id}/status`, {
    method: 'PATCH',
    body: JSON.stringify({ status }),
  });
}
